From ee966c6f74580d915549d2c7ae09e21e23220fa9 Mon Sep 17 00:00:00 2001
From: Zhenyong Chen <b07273@freescale.com>
Date: Tue, 30 Dec 2014 01:27:02 +0800
Subject: [PATCH 0766/1594] MGS-386 [#1506] Power mutex track is buggy in
 gckKERNEL_Dispatch

commit 6713e2ef0cd5fc98456635758823932e025b3c37 from
git://git.freescale.com/imx/linux-2.6-imx.git

In function gckKERNEL_Dispatch, powerMutexAcquired is used to track
power mutex. It is missing from gcvHAL_WRITE_REGISTER. And in some
places return value of gckOS_AcquireMutex is not checked.

Date: Dec 29, 2014
Signed-off-by: Zhenyong Chen <b07273@freescale.com>
Acked-by: Jason Liu
---
 drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
index b14708d..4f1ae0a 100644
--- a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
+++ b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
@@ -2045,7 +2045,7 @@ gckKERNEL_Dispatch(
             gctUINT32 coreId = 0;
             gctUINT32 coreSelect = Interface->u.ReadRegisterDataEx.coreSelect;
 
-            gckOS_AcquireMutex(Kernel->os, Kernel->hardware->powerMutex, gcvINFINITE);
+            gcmkONERROR(gckOS_AcquireMutex(Kernel->os, Kernel->hardware->powerMutex, gcvINFINITE));
             powerMutexAcquired = gcvTRUE;
             gcmkONERROR(gckHARDWARE_QueryPowerManagementState(Kernel->hardware,
                     &power));
@@ -2142,7 +2142,8 @@ gckKERNEL_Dispatch(
         {
             gceCHIPPOWERSTATE power;
 
-            gckOS_AcquireMutex(Kernel->os, Kernel->hardware->powerMutex, gcvINFINITE);
+            gcmkONERROR(gckOS_AcquireMutex(Kernel->os, Kernel->hardware->powerMutex, gcvINFINITE));
+            powerMutexAcquired = gcvTRUE;
             gcmkONERROR(gckHARDWARE_QueryPowerManagementState(Kernel->hardware,
                                                                   &power));
             if (power == gcvPOWER_ON)
@@ -2161,6 +2162,7 @@ gckKERNEL_Dispatch(
                 status = gcvSTATUS_CHIP_NOT_READY;
             }
             gcmkONERROR(gckOS_ReleaseMutex(Kernel->os, Kernel->hardware->powerMutex));
+            powerMutexAcquired = gcvFALSE;
         }
 #else
         /* No access from user land to write registers. */
-- 
1.7.5.4

